### ISSUE TITLE
Compiling Gnu Make

### ISSUE THREAD

**Issue by andrewbird** [tkchia/libi86#2]
Hi I'm looking to compile Gnu Make for DOS using ia16-elf-gcc as I want a non DPMI version of it. I'm running into trouble because of missing functions from `process.h` like `spawnvpe()`. It might be possible to work around this I guess, but before I invest a lot of time into this, can you see any fundamental problems that you think I would face with the overall goal?

**Comment by tkchia** [comment]
Hello @andrewbird,

One potential problem I can think of is that `ia16-elf-gcc` currently lacks support for far-data memory models (compact, large, huge) --- and I am not sure if this will have an adverse impact on GNU `make`'s capabilities.

I see that the GNUish Project's port of GNU `make` 3.71 (see [`gmake371.zip`](http://www.retroarchive.org/garbo/pc/programming/)) was built with the large memory model.

Well... I guess one thing you can try is to create stubs for `spawnvpe(`...`)` and friends, and see whether `make` compiled with `ia16-elf-gcc` has any problem parsing and processing makefiles first.

Thank you!

**Comment by andrewbird** [comment]
Thanks for your insight and the great compiler! I'm probably going to pause on this quest for now.

Thank you!

**Comment by tkchia** [comment]
Hello @andrewbird,

> Thanks for your insight and the great compiler! I'm probably going to pause on this quest for now.

No problem. :-)

Also, may I ask why you were looking to build a non-DPMI version of GNU `make`?  I had also been contemplating building a 16-bit version of either GNU `make` or Vadura's `dmake`, but I was not sure of the usefulness of such a thing.

Thank you!

**Comment by andrewbird** [comment]
Hello @tkchia,
   I've been wanting to tidy up the build process for PC-MOS, see https://github.com/andrewbird/pcmos386/commit/022d105e91704b8411a96f8be5b79dea91e9da79 for my work in progress. There are some files to be built that are named `$$abc.sys` etc, which are awkward to handle (in different ways, and not consistently within the tool) with each of the proprietary Makes that are currently part of the repository. Gnu Make does handle these consistently, only needing some simple escaping of the dollar signs. The nice automatic variables and string functions really help to simplify the code too. Of course switching to Open Source tools is always a great goal. However I see two problems with using Gnu Make at the moment though; firstly the repository needs to be buildable by others, some of whom use real hardware, so as it's a DPMI program I also have to include `cwsdmpi` into the repository; secondly I noticed it was about half the speed of the Microsoft Make tool currently used. I did read yesterday an article somewhere on the Internet about optimising programs for DPMI, and it suggested that as a general rule what really slows them down are switches between real and protected mode and that file I/O etc should be chunked together to minimise the penalty. With a Make program it's pretty much all file I/O so I concluded that maybe DPMI is not ideal for this use.
  At the time I was having a different problem, which @ stsp later fixed for me, and I wrongly attributed it to an interaction between Gnu Make with DPMI, FDPP/Comcom32 and the need to use loadfix with Microsoft C 5.10. My question to yourself yesterday arose in my attempt to exclude DPMI from that combination. That specific issue is now fixed, so there's no pressing need for a non DPMI version, but if one were available I'd gladly switch to it.

Thank you!
